### ISSUE TITLE
App is Crashing when i use flutter fragment

### ISSUE THREAD

**Issue by iamaravindh123** [flutter/flutter#50950]
## Steps to Reproduce

1.  Import   flutter module in existing android app

2. create fragement = FlutterFragment.withCachedEngine("shop").shouldAttachEngineToActivity(false)
                .renderMode(FlutterView.RenderMode.texture).build()
3.  navigate to new activity, app will crash 

**Expected results:**  surface cant be null,while calling on stop

**Actual results:** the surface is null . while calling on stop 

** ps- happening in some of the phones especially android 21,23 version tested with galaxy s6 is crashing, pixel 2 it is working

  <summary>Crash in Android</summary>

```java
    java.lang.RuntimeException: Unable to stop activity {com.shopup.reseller/com.shopup.reseller.HomeActivity}: java.lang.NullPointerException: Attempt to invoke virtual method 'void android.view.Surface.release()' on a null object reference
        at android.app.ActivityThread.performStopActivityInner(ActivityThread.java:3344)
        at android.app.ActivityThread.handleStopActivity(ActivityThread.java:3390)
        at android.app.ActivityThread.access$1100(ActivityThread.java:144)
        at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1307)
        at android.os.Handler.dispatchMessage(Handler.java:102)
        at android.os.Looper.loop(Looper.java:135)
        at android.app.ActivityThread.main(ActivityThread.java:5221)
        at java.lang.reflect.Method.invoke(Native Method)
        at java.lang.reflect.Method.invoke(Method.java:372)
        at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:899)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:694)
     Caused by: java.lang.NullPointerException: Attempt to invoke virtual method 'void android.view.Surface.release()' on a null object reference
        at io.flutter.embedding.android.FlutterTextureView.disconnectSurfaceFromRenderer(FlutterTextureView.java:198)
```

Flutter  doctor -v logs
 
<details>

```bash
[✓] Flutter (Channel beta, v1.14.6, on Mac OS X 10.15.2 19C57, locale en-IN)
    • Flutter version 1.14.6 at /Users/asthme/sdks/flutter
    • Framework revision fabeb2a16f (3 weeks ago), 2020-01-28 07:56:51 -0800
    • Engine revision c4229bfbba
    • Dart version 2.8.0 (build 2.8.0-dev.5.0 fc3af737c7)

 
[✓] Android toolchain - develop for Android devices (Android SDK version 29.0.2)
    • Android SDK at /Users/asthme/Library/Android/sdk
    • Android NDK location not configured (optional; useful for native profiling support)
    • Platform android-29, build-tools 29.0.2
    • Java binary at: /Applications/Android Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 1.8.0_202-release-1483-b49-5587405)
    • All Android licenses accepted.

[✗] Xcode - develop for iOS and macOS
    ✗ Xcode installation is incomplete; a full installation is necessary for iOS development.
      Download at: https://developer.apple.com/xcode/download/
      Or install Xcode via the App Store.
      Once installed, run:
        sudo xcode-select --switch /Applications/Xcode.app/Contents/Developer
        sudo xcodebuild -runFirstLaunch
    ✗ CocoaPods not installed.
        CocoaPods is used to retrieve the iOS and macOS platform side's plugin code that responds to your plugin usage on the Dart side.
        Without CocoaPods, plugins will not work on iOS or macOS.
        For more info, see https://flutter.dev/platform-plugins
      To install:
        sudo gem install cocoapods

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 3.5)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin version 43.0.1
    • Dart plugin version 191.8593
    • Java version OpenJDK Runtime Environment (build 1.8.0_202-release-1483-b49-5587405)

[✓] Connected device (4 available)
    • ASUS I01WD        • K6AIHM00A1615LZ     • android-arm64  • Android 10 (API 29)
    • Samsung Galaxy S5 • 192.168.56.102:5555 • android-x86    • Android 4.4.4 (API 19)
    • Chrome            • chrome              • web-javascript • Google Chrome 79.0.3945.130
    • Web Server        • web-server          • web-javascript • Flutter Tools
````

```bash
 [/Users/asthme/sdks/flutter/] git -c log.showSignature=false log -n 1 --pretty=format:%H
[  +31 ms] Exit code 0 from: git -c log.showSignature=false log -n 1 --pretty=format:%H
[        ] fabeb2a16f1d008ab8230f450c49141d35669798
[        ] executing: [/Users/asthme/sdks/flutter/] git describe --match v*.*.* --first-parent --long --tags
[   +9 ms] Exit code 0 from: git describe --match v*.*.* --first-parent --long --tags
[        ] v1.14.6-0-gfabeb2a16
[   +5 ms] executing: [/Users/asthme/sdks/flutter/] git rev-parse --abbrev-ref --symbolic @{u}
[   +9 ms] Exit code 0 from: git rev-parse --abbrev-ref --symbolic @{u}
[        ] origin/beta
[        ] executing: [/Users/asthme/sdks/flutter/] git ls-remote --get-url origin
[   +7 ms] Exit code 0 from: git ls-remote --get-url origin
[        ] https://github.com/flutter/flutter.git
[  +40 ms] executing: [/Users/asthme/sdks/flutter/] git rev-parse --abbrev-ref HEAD
[   +8 ms] Exit code 0 from: git rev-parse --abbrev-ref HEAD
[        ] beta
[   +4 ms] executing: sw_vers -productName
[  +12 ms] Exit code 0 from: sw_vers -productName
[        ] Mac OS X
[        ] executing: sw_vers -productVersion
[  +12 ms] Exit code 0 from: sw_vers -productVersion
[        ] 10.15.2
[        ] executing: sw_vers -buildVersion
[  +12 ms] Exit code 0 from: sw_vers -buildVersion
[        ] 19C57
[  +24 ms] Artifact Instance of 'AndroidMavenArtifacts' is not required, skipping update.
[        ] Artifact Instance of 'AndroidGenSnapshotArtifacts' is not required, skipping update.
[        ] Artifact Instance of 'AndroidInternalBuildArtifacts' is not required, skipping update.
[        ] Artifact Instance of 'IOSEngineArtifacts' is not required, skipping update.
[        ] Artifact Instance of 'FlutterWebSdk' is not required, skipping update.
[   +2 ms] Artifact Instance of 'WindowsEngineArtifacts' is not required, skipping update.
[        ] Artifact Instance of 'MacOSEngineArtifacts' is not required, skipping update.
[        ] Artifact Instance of 'LinuxEngineArtifacts' is not required, skipping update.
[        ] Artifact Instance of 'LinuxFuchsiaSDKArtifacts' is not required, skipping update.
[        ] Artifact Instance of 'MacOSFuchsiaSDKArtifacts' is not required, skipping update.
[        ] Artifact Instance of 'FlutterRunnerSDKArtifacts' is not required, skipping update.
[        ] Artifact Instance of 'FlutterRunnerDebugSymbols' is not required, skipping update.
[  +23 ms] executing: /Users/asthme/Library/Android/sdk/platform-tools/adb devices -l
[   +6 ms] Exit code 0 from: /Users/asthme/Library/Android/sdk/platform-tools/adb devices -l
[        ] List of devices attached
           K6AIHM00A1615LZ        device usb:339738624X product:WW_I01WD model:ASUS_I01WD device:ASUS_I01WD transport_id:2
           192.168.56.102:5555    device product:vbox86p model:Samsung_Galaxy_S5 device:vbox86p transport_id:3
[  +20 ms] executing: /Users/asthme/sdks/flutter/bin/cache/artifacts/libimobiledevice/idevice_id -h
[  +43 ms] executing: /usr/bin/xcode-select --print-path
[   +9 ms] Exit code 0 from: /usr/bin/xcode-select --print-path
[        ] /Library/Developer/CommandLineTools
[        ] executing: /usr/bin/xcodebuild -version
[   +2 ms] Exit code 1 from: /usr/bin/xcodebuild -version
[        ] xcode-select: error: tool 'xcodebuild' requires Xcode, but active developer directory '/Library/Developer/CommandLineTools' is a command line tools instance
[   +7 ms] /Users/asthme/Library/Android/sdk/platform-tools/adb -s K6AIHM00A1615LZ shell getprop
[  +85 ms] /Users/asthme/Library/Android/sdk/platform-tools/adb -s 192.168.56.102:5555 shell getprop
[  +12 ms] Artifact Instance of 'AndroidMavenArtifacts' is not required, skipping update.
[   +2 ms] Artifact Instance of 'AndroidInternalBuildArtifacts' is not required, skipping update.
[        ] Artifact Instance of 'IOSEngineArtifacts' is not required, skipping update.
[   +1 ms] Artifact Instance of 'WindowsEngineArtifacts' is not required, skipping update.
[        ] Artifact Instance of 'MacOSEngineArtifacts' is not required, skipping update.
[        ] Artifact Instance of 'LinuxEngineArtifacts' is not required, skipping update.
[        ] Artifact Instance of 'LinuxFuchsiaSDKArtifacts' is not required, skipping update.
[        ] Artifact Instance of 'MacOSFuchsiaSDKArtifacts' is not required, skipping update.
[        ] Artifact Instance of 'FlutterRunnerSDKArtifacts' is not required, skipping update.
[        ] Artifact Instance of 'FlutterRunnerDebugSymbols' is not required, skipping update.
[   +7 ms] executing: /usr/bin/xcodebuild -version
[   +3 ms] Exit code 1 from: /usr/bin/xcodebuild -version
[        ] xcode-select: error: tool 'xcodebuild' requires Xcode, but active developer directory '/Library/Developer/CommandLineTools' is a command line tools instance
[  +42 ms] More than one device connected; please specify a device with the '-d <deviceId>' flag, or use '-d all' to act on all devices.
[   +3 ms] ro.hardware = qcom
[        ] ro.hardware = vbox86
[        ] ro.build.characteristics = default
[ +111 ms] ASUS I01WD        • K6AIHM00A1615LZ     • android-arm64  • Android 10 (API 29)
[        ] Samsung Galaxy S5 • 192.168.56.102:5555 • android-x86    • Android 4.4.4 (API 19)
[        ] Chrome            • chrome              • web-javascript • Google Chrome 79.0.3945.130
[        ] Web Server        • web-server          • web-javascript • Flutter Tools
```

</details>

**Comment by TahaTesser** [comment]
Hi @iamaravindh123 
can you please provide your `flutter doctor -v` and `flutter run --verbose`? 
Thank you

**Comment by iamaravindh123** [comment]
@TahaTesser updated ...

**Comment by allenyulun** [comment]
in galaxy s6 , surfaceTextureListener onSurfaceTextureDestroyed called first， FlutterTextureView.disconnectSurfaceFromRenderer called twice  #https://github.com/flutter/engine/pull/18295

**Comment by pengrad** [comment]
We are also experiencing this crash and have to downgrade to 1.12 flutter version.
This fix flutter/engine#18295 will help.

@chinmaygarde  any estimates when this could be released?

@allenyulun could you please sign CLA in PR? :) It probably won't be merged without that.

**Comment by lishuliang28** [comment]
The page of the low version page is closed and the app crashes. I see that some people have mentioned it in previous versions, but the latest version of v1.20.1 still exists. When can this problem be upgraded to a stable version? Or we can't use it

**Comment by pengrad** [comment]
This fix https://github.com/flutter/engine/pull/19501 is definitely not in v1.20.1 release, but 1.21 will have it (not sure when it will be stable though)

I have some workaround:
1) Make own fragment which extends `FlutterFragment`
2) Create your fragment as said in [CachedEngineFragmentBuilder javadoc](https://api.flutter.dev/javadoc/io/flutter/embedding/android/FlutterFragment.CachedEngineFragmentBuilder.html)
3) override onDestroyView and wrap super with try-catch
```kotlin
override fun onDestroyView() {
    try {
        super.onDestroyView()
    } catch (e: Exception) {
        // known issue of the flutter engine
        // https://github.com/flutter/flutter/issues/58767
        // remove when these changes are in stable engine
        // https://github.com/flutter/engine/pull/19501
    }
}
```

**Comment by arsenioo** [comment]
> I have some workaround: fot 
I doubt this is a workaround, I should better say it is just a way to bury the problem even deeper

**Comment by pengrad** [comment]
> I have some workaround: fot
> I doubt this is a workaround, I should better say it is just a way to bury the problem even deeper

it works fine in production catching thousands of crashes, I'm logging catched exception in crashlytics.
alternative is to use 1.12 flutter engine or wait for 1.23 in November (?)
both 1.17 and 1.20 will crash here

**Comment by TahaTesser** [comment]
Hi @iamaravindh123 @pengrad  

I cannot reproduce this using latest `master` channel 

```kt
mport android.os.Bundle
import androidx.fragment.app.FragmentActivity
import androidx.fragment.app.FragmentManager
import io.flutter.embedding.android.FlutterFragment
import io.flutter.embedding.android.RenderMode

class CachedEngineActivity : FragmentActivity() {

    companion object {
        private const val TAG_FLUTTER_FRAGMENT = "flutter_fragment"

    }

    private var flutterFragment: FlutterFragment? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.cached_activity)

        val fragmentManager: FragmentManager = supportFragmentManager

        flutterFragment = fragmentManager
            .findFragmentByTag(TAG_FLUTTER_FRAGMENT) as FlutterFragment?

        if (flutterFragment == null) {
//            var cachedFlutterFragment = FlutterFragment.withCachedEngine(ENGINE_ID).build<FlutterFragment>()
            var cachedFlutterFragment =
                FlutterFragment.withCachedEngine(ENGINE_ID).shouldAttachEngineToActivity(false)
                    .renderMode(RenderMode.image).build<FlutterFragment>()


            fragmentManager
                .beginTransaction()
                .add(
                    R.id.fragment_container_cached,
                    cachedFlutterFragment,
                    TAG_FLUTTER_FRAGMENT
                )
                .commit()
        }
    }
}
```

Can you please provide your `flutter doctor -v`  and a complete reproducible minimal code sample in are
Thank you

**Comment by pengrad** [comment]
Hi @TahaTesser 
it was fixed by this PR https://github.com/flutter/engine/pull/19501
it just wan't available in stable release until 1.23 version. 
Now it's included in stable, this issue can be closed.

**Comment by TahaTesser** [comment]
@pengrad 
Thanks for confirming, I feel safe to close this issue
Thanks everyone
